zerver/lib/import_realm.py

```diff
@@ -345,7 +346,7 @@ def import_uploads_s3(bucket_name: str, import_dir: Path, processing_avatars: bo
                 sanitize_name(os.path.basename(record['path']))
             ])
             key.key = s3_file_name
-            path_maps['attachment_path'][record['path']] = s3_file_name
+            path_maps['attachment_path'][record['s3_path']] = s3_file_name
```

Comment:
Can you explain this change in more detail?  I don't understand it.  Also, I think this commit:
```
    import: Fix S3 backend upload bugs.
    
    'import_uploads_s3' and 'import_uploads_local' should follow
    the same overall pattern for functioning. Here I have fixed
    certain issues like:
    
    * 'processing_emojis' should have an 'elif' condition here as
      we want the function to work for either avatars, emojis or
      uploads at one time.
    * 'import_uploads_s3' should be passed with the parameter
      'processing_emojis' from 'import_uploads'.
    * For the S3 backend uploads, 'attachment_path' should be
      saved with the 's3_path' of the record.
    * All the avatars should be processed later on to run the
      'ensure_medium_avatar_image' function. This is similar to
      'import_uploads_local'.
```

Should be 4 commits: one for each issue.  There's no reason to have 4 basically independent changes squashed together like that.

Reply:
I'll break this into different commits and update it.

Reply:
While saving the records for the uploads, we edit the `path` and save it. ([reference](https://github.com/zulip/zulip/blob/master/zerver/lib/export.py#L1045-L1049)). 
Here in `import_realm`, we use `path_maps['attachment_path']` to ensure that the `path_id` in `Attachment` is updated ([reference](https://github.com/zulip/zulip/blob/master/zerver/lib/import_realm.py#L682-L683)). So the original `attachment['path_id']` referes to `s3_path` and not `path` as `path` is changed [here](https://github.com/zulip/zulip/blob/master/zerver/lib/export.py#L1045-L1049).

Hence if we have `path_maps['attachment_path'][record['path']] = s3_file_name` above, we get a `KeyError` [here](https://github.com/zulip/zulip/blob/master/zerver/lib/import_realm.py#L682-L683) (as original `attachment['path_id']` _is not equal_ to `record['path']`). Does this make sense?

Reply:
Yep; let's put a piece of that explanation in a comment